src/Dsll.hs

```diff
@@ -0,0 +1,322 @@
+--   operations followed by a collection of predicates
+dsllProg :: Parser DSLLProg
+dsllProg = do
+    listCd <- cdParser `sepEndBy` newline'
+    listVd <- vdParser `sepEndBy` newline'
+    listOd <- odParser `sepEndBy` newline'
+    listPd <- pdParser `sepEndBy` newline'
+    return DSLLProg { cd = listCd, vd = listVd, od = listOd, pd = listPd }
+
+-- | type constructor parser
+cdParser, cd1, cd2 :: Parser Cd
+cdParser = try cd1 <|> cd2
+cd1 = do
+    rword "tconstructor"
+    name <- identifier
+    (y, k) <- parens ykParser
+    colon
+    t' <- typeParser
+    pos <- getPosition
+    return Cd { nameCd = name, inputCd = zip y k, outputCd = t' }
+cd2 = do
+    rword "tconstructor"
+    name <- identifier
+    colon
+    t' <- typeParser
+    return Cd { nameCd = name, inputCd = [], outputCd = t' }
+
+-- | parser for the (y,k) list
+ykParser :: Parser ([Y], [K])
+ykParser = unzip <$> (yWithKind `sepBy1` comma)
+    where yWithKind = (,) <$> (yParser <* colon) <*> kParser
+
+-- | parser for the (b,t) list
+xtParser :: Parser ([Var], [T])
+xtParser = unzip <$> (varWithType `sepBy1` comma)
+    where varWithType = (,) <$> (varParser <* colon) <*> tParser
+
+-- | parser for the (x, Prop) list
+xPropParser :: Parser ([Var], [Prop])
+xPropParser = unzip <$> (vpPair `sepBy1` comma)
+    where vpPair = (,) <$> (varParser <* colon) <*> propParser
+
+-- | var constructor parser
+vdParser :: Parser Vd
+vdParser = do
+  rword "vconstructor"
+  name <- identifier
+  (y', k') <- option ([], []) ykParser
+  (b', t') <- option ([], []) xtParser
+  colon
+  t'' <- tParser
+  return Vd { nameVd = name, varsVd = (zip y' k'), typesVd = (zip b' t'), toVd = t'' }
+
+  -- | operation parser
+odParser :: Parser Od
+odParser = do
+  rword "operator"
+  name <- identifier
+  (y', k') <- option ([], []) $ brackets ykParser
+  (b', t') <- option ([], []) $ parens   xtParser
+  colon
+  t'' <- tParser
+  return Od { nameOd = name, varsOd = (zip y' k'), typesOd = (zip b' t'), toOd = t'' }
+
+-- | predicate parser
+pdParser, pd1, pd2 :: Parser Pd
+pdParser = try pd1 <|> pd2
+pd1 = do
+  rword "predicate"
+  name <- identifier
+  (y', k') <- option ([], []) $ brackets ykParser
+  (b', t') <- option ([], []) $ parens   xtParser
+  colon
+  p' <- propParser
+  return (Pd1Const (Pd1 { namePd1 = name, varsPd1 = (zip y' k'), typesPd1 = (zip b' t'), toPd1 = p' }))
+pd2 = do
+  rword "predicate"
+  name <- identifier
+  (b', prop') <- parens xPropParser
+  colon
+  p' <- propParser
+  return (Pd2Const (Pd2 { namePd2 = name, propsPd2 = (zip b' prop'), toPd2 = p' }))
+
+--------------------------------------- DSLL Semantic Checker ---------------------------
+
+-- | 'check' is the top-level semantic checking function. It takes a DSLL
+-- program as the input, checks the validity of the program acoording to the typechecking rules, and outputs
+-- a collection of information.
+check :: DSLLProg -> VarEnv
+check p =  let env1  = foldl checkTypeConstructors initE (cd p)
```

Comment:
check if there are multiple declarations of the same name (tycon, valcon, ops, preds, or across any of those) -- and throw an error if so